docs/v1/P4Runtime-Spec.mdk

```diff
@@ -4550,7 +4566,11 @@ The egress packets may be distinguished for further processing in the egress
 using the `instance` metadata. Note that a packet may not be both unicast and
 multicast; if the multicast group is set, it will override the unicast egress
 port. If the P4 `multicast_group` metadata is set to a value that is not
-programmed in the PRE, then the packet is dropped.
+programmed in the PRE, then the packet is dropped. For each replicas(i) in
+the `multicast_group_entry`, there must be an associated backups(i) or
+`multicast_group_entry` cannot have any backups at all. Fallback ports for
+multicast uses one of the backups(i) replicas when the primary `port` of a
+replicas(i) becomes unavailable.
```

Comment:
So the intent is that there must be either 0 backups in a multicast_group_entry, or there must be exactly n backups, where n is the number of repliaces in the multicast_group_entry?

Did you consider instead adding the backups message as a sub-message of the Replica message?  Then the association between the primary and backups would be more direct, and you could even make the backups optional on a per-replica basis, e.g. a mix of some replicas that have backups, and some that do not.

Reply:
The intent is as you mentioned where we either have 0 backups or exactly n backups, where n is the number of replicas in multicast_group_entry.

There were considerations for other options. If we went with something like this where we made the Replica own multiple other Replicas, it provides an nesting behavior that would be undesired.
```
message Replica {
	oneof port_kind {
		uint32 egress_port = 1 [deprecated=true];
		bytes port = 3;
        }
        uint32 instance = 2;
        repeated Replica fallback_replicas = 4;
}
```

Similarly if we chose to create another message FallbackReplicas and keep it nested in Replica, it would create a potential syncing issue between FallbackReplica and Replica if changes were made.
```
message FallbackReplica {
	bytes port = 1;
	uint32 instance = 2;
}

message Replica {
	oneof port_kind {
		uint32 egress_port = 1 [deprecated=true];
		bytes port = 3;
         }
         uint32 instance = 2;
         repeated FallbackReplica fallback_replicas = 4;
}
```

Reply:
That's a fair analogy. I got some clarification from the ASIC team.

TLDR; allowing `MulticastGroupEntry` to own the `FallbackReplicas` is a 1:1 ASIC representation and results in fewer ASIC operations. 

However, I feel as though the increased flexibility of having Replicas owning the backups is desired. I've updated my PR accordingly. Will mark this comment as resolved since it is no longer relevant